src/provider/view/components/AppCard.tsx

```diff
@@ -17,10 +17,11 @@ export function AppCard(props: AppCardProps): React.ReactElement {
         event.stopPropagation();
         openHandler(app);
     };
+
     return (
-        <li className="app-card" onClick={clickHandler}>
+        <li id={app.name} className="app-card" onClick={clickHandler}>
```

Comment:
Will `app.name` always be a valid CSS selector string? (not contain invalid characters e.g. spaces)

Reply:
Possibly not, but I think that's ok?

The ID is here for allowing tests to press the button for given test apps, so we're controlling the app IDs in this case. Where we're not controlling app IDs, given this is going through react rather than going directly into HTML, even if we try doing something nasty in the ID, like trying to close the quote marks surrounding the id and inject some HTML, everything still just works (including getElementById).

What it would break is if we were trying to style buttons by app name, but I wouldn't worry about that while we're not doing this.

If we definitely want to escape this, I think the most sensible thing to be would be to use `encodeURI`, but even then we might not have a valid ID, if it starts with a digit.

Reply:
Ok, so, having spent more time digging into this rabbit hole than I probably should do, turns out this _is_ a rabbit hole.

The robust thing to do would be to
- Leave AppCard.tsx as is
- Import this - https://github.com/mathiasbynens/cssesc
- Use rewrite our query selector in `raiseIntent` as:
    `#${cssesc(appName, {isIdentifier: true})}`

and this then work no matter what's in the appName string - spaces, new lines, emoji...

But, I feel like this isn't work adding an extra dependency to solve

Reply:
This feels like one of those things that will come back as a bug in 3-6 months time once we've all forgotten about it. Is there not some better unique identifier we could use that would avoid needing the app name, or could we instead just come up with some deterministic mapping from app name to a selector-safe string, like a very simple hash or something.

Also, do we know what happens in the UI if an invalid string is passed in as an app name? I imagine it could cause weird things to the DOM depending on what it is, but maybe react can escape it properly to avoid that.

Reply:
Having spent some time trying to break this, either react is escaping whatever is put in here, or behind the scenes, it's just doing `element.id = appName`, so escaping is a non-issue here

Reply:
Ok, that's fair. If it's only going to be a problem for the query selector in the test util then I'm less concerned as we have complete control over that env.

Reply:
Yeah I guess it isn't that important at the moment.
This is how I would probably do it:
`<li className="app-card" data-appName={app.name}>...`

The selector would then be:
`.app-card [data-appName="test-app-1"]`

Reply:
Oooooooh. Actually yeah, that's really nice. I'll leave the escaping, but I'll make the change so it's using `data-appName`

Reply:
From memory that's how we do it in notifications and it works well.